File: exercism/python/pig-latin/pig_latin.py
from itertools import chain, repeat


CONSONANT = 'c'
VOWEL = 'v'
RULES = tuple(chain(
    zip(('ch', 'qu', 'squ', 'thr', 'th', 'sch'), repeat(CONSONANT)),
    zip(('yt', 'xr'), repeat(VOWEL)),
    zip('bcdfghjklmnpqrstvwxyz', repeat(CONSONANT)),
    zip('aeiou', repeat(VOWEL))
))


def translate(phrase):
    return ' '.join(map(translate_word, phrase.split(' ')))


def translate_word(word):
    head, tail = sound_split(word)
    return '{}{}ay'.format(head, tail)


def sound_split(word):
    for stem, case in RULES:
        head, tail = word[:len(stem)], word[len(stem):]
        if head == stem:
            return (head, tail) if case == VOWEL else (tail, head)

File: exercism/javascript/pig-latin/pig-latin.js
const fp = require('lodash/fp')

const translate = (phrase) =>
  phrase.split(' ').map(translateWord).join(' ')

const translateWord = (word) =>
  splitOnSound(word).join('') + 'ay'

const consonant = Symbol('c')
const vowel = Symbol('v')
const as = (y, xs) => xs.map(x => [y, x])
const rules = fp.flatten([
  as(consonant, ['ch', 'qu', 'squ', 'thr', 'th', 'sch']),
  as(vowel, ['yt', 'xr']),
  as(consonant, 'bcdfghjklmnpqrstvwxyz'.split('')),
  as(vowel, 'aeiou'.split(''))
])

const splitOnSound = (word) =>
  fp.flow(
    fp.filter(ruleFits(word)),
    fp.map(partOrder(word)),
    fp.compact,
    fp.head
  )(rules)

const ruleFits = fp.curry((word, [_, stem]) =>
  stem === word.slice(0, stem.length))

const partOrder = fp.curry((word, [type, stem]) =>
  (type === vowel ? fp.identity : fp.reverse)(
    [stem, word.slice(stem.length)]))

module.exports = { translate }
